conjure-lib/src/main/java/com/palantir/conjure/java/lib/internal/ConjureDoubleList.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.lib.internal;
+
+import com.palantir.logsafe.exceptions.SafeUnsupportedOperationException;
+import java.util.AbstractList;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.RandomAccess;
+import org.eclipse.collections.impl.list.mutable.primitive.DoubleArrayList;
+
+public final class ConjureDoubleList extends AbstractList<Double> implements RandomAccess {
+    private final DoubleArrayList delegate;
+
+    public ConjureDoubleList() {
+        this.delegate = new DoubleArrayList();
+    }
+
+    public ConjureDoubleList(Iterable<Double> iterable) {
+        if (iterable instanceof Collection) {
+            this.delegate = new DoubleArrayList(((Collection<Double>) iterable).size());
+        } else {
+            this.delegate = new DoubleArrayList();
+        }
+        for (double e : iterable) {
+            delegate.add(e);
+        }
+    }
+
+    @Override
+    public int size() {
+        return delegate.size();
+    }
+
+    @Override
+    public void add(int index, Double toAdd) {
+        delegate.addAtIndex(index, toAdd);
+    }
+
+    @Override
+    public Double get(int index) {
+        Objects.checkIndex(index, delegate.size());
+        return delegate.get(index);
+    }
+
+    @Override
+    public boolean addAll(int _index, Collection<? extends Double> _collection) {
+        throw new SafeUnsupportedOperationException("This operation is unsupported");
```

Comment:
This is also weird. The API for eclipse collections provides `boolean addAllAtIndex(int var1, DoubleIterable var2)`, but LazyDoubleIterable is in the api package which could cause classloader bug in java < 21. There are workarounds, like implementing our own DoubleIterable, but given this method is never accessed, I left the exception here.

Reply:
Why are we providing our own implementations instead of using eclipse-collections types directly?

Reply:
See this https://github.com/palantir/conjure-java/pull/2307#discussion_r1597703482

Reply:
I'm not sure I understand how `AbstractList` is relevant, though

Reply:
```
public static List<Double> newNonNullDoubleList() {
    return new DoubleArrayList();
}
```

won't work because DoubleArrayList doesn't extend AbstractList<Double> or implement List<Double>

Reply:
I don't think it needs to -- shouldn't we implement the collection factories using something along these lines in order to expose eclipse-collections using standard java collection facades?

```java
public static List<Double> newNonNullDoubleList() {
    return new BoxedMutableDoubleList(new DoubleArrayList());
}
```

Reply:
Whats the benefit of this approach versus using a delegate instantiated in the parent's constructor? As I understand these all will do the same thing?

Reply:
`return new BoxedMutableDoubleList(new DoubleArrayList());` uses classes that we don't have to build and maintain ourselves

Reply:
Where do you see BoxedMutableDoubleList? I can't find it, so I assumed this would be what ConjureDoubleList.java is right now

Reply:
Seems like that isn't functionality provided by eclipse-collections yet https://github.com/eclipse/eclipse-collections/issues/1408

Reply:
Ah it is implemented on the tip of the development branch, but has not yet merged

Reply:
oof that makes sense y i was confused. I was looking at their 11.1 release. Last release was in 2022 so idk when this will come out. Keep current pr as is and bump version when its out?

Reply:
I think that’s reasonable, I think a couple minor tweaks will make things a bit easier:
1. Let’s javadoc the plan to use eclipse-collections 12 wrappers, since we don’t know how far out that will be
2. Let’s make our classes similar in shape/structure to the eclipse-collections api so that the transition won’t require much code change aside from deleting our implementations.
3. Our collection implementations should be package private, and only created (without exposing type information) through the ConjureCollections factories

Reply:
done https://github.com/palantir/conjure-java/pull/2307/commits/e8c2f78e85390ea9435c427d5ae7db2b0aa484ef

eclipse-collections implementation for reference: https://github.com/eclipse/eclipse-collections/blob/master/eclipse-collections-code-generator/src/main/resources/impl/list/mutable/boxedPrimitiveList.stg